postgres-appliance/bootstrap/clone_with_basebackup.py

```diff
@@ -0,0 +1,119 @@
+    args = parser.parse_args()
+
+    options = namedtuple('Options', 'name datadir bindir pgpassfile')
+
+    result=options(name=args.scope, datadir=args.datadir, bindir=args.bindir, pgpassfile=args.from_pgpass)
+    return result
+
+def parse_pgpass_file(pgpassfile):
+    with open(pgpassfile, 'r') as f:
+        line = f.readline()
+        return parse_pgpass_line(line)
+
+
+def parse_pgpass_line(line):
+    """ parses a single pgpass line and returns a connection string (without the password)
+
+        >>> parse_pgpass_line('127.0.0.1:5432:db:user:pass')
+        pgpass(host='127.0.0.1', port='5432', dbname='db', user='user')
+        >>> parse_pgpass_line('\:\:1:5432:"ba\\\\\\\\z":qiz:foo')
+        pgpass(host='::1', port='5432', dbname='"ba\\\\z"', user='qiz')
+        >>> parse_pgpass_line('127.0.0.1:1234:db:user:')
+        Traceback (most recent call last):
+            ...
+        Exception: pgpass file contains an empty field
+        >>> parse_pgpass_line('127\.0.0.1:1234:db:user:')
+        Traceback (most recent call last):
+            ...
+        Exception: pgpass file has unescaped '\\' character
+    """
+
+    fields = []
+    escape = False
+    current = []
+    for c in line:
+        if c == ':' and not escape:
+            if len(current) == 0:
+                raise Exception("pgpass file contains an empty field")
+            fields.append(''.join(current))
+            current = []
+            continue
+        if c == '\\' and not escape:
+            escape = True
+            continue
+        if escape:
+            if c not in (':', '\\'):
+                raise Exception("pgpass file has unescaped '\\' character")
+            escape = False
+        current.append(c)
+    # process the last field in line
+    if escape:
+        raise Exception("pgpass file has unescaped '\\' character")
+    if len(current) == 0:
+        raise Exception("pgpass file contains an empty field")
+    fields.append(''.join(current))
+    if len(fields) != 5:
+        raise Exception("pgpass file must be in a format 'hostname:port:database:username:password'")
+    return namedtuple('pgpass', 'host port dbname user')._make(fields[:4])
+
+def escape_value(val):
+    quote = False
+    temp = []
+    for c in val:
+        if c.isspace():
+            quote = True
+        elif c in ('\'', '\\'):
+            temp.append('\\')
+        temp.append(c)
+    result = ''.join(temp)
+    return result if not quote else '\'{0}\''.format(result)
+
+def prepare_connection(options):
+    pgpass = parse_pgpass_file(options.pgpassfile)
+    connection = []
+    for attname in ('host', 'port', 'user', 'dbname'):
+        attvalue = getattr(pgpass, attname)
+        if attvalue and attvalue != '*':
+            connection.append('{0}={1}'.format(attname, escape_value(attvalue)))
+
+    return ' '.join(connection), {'PGPASSFILE': options.pgpassfile}
```

Comment:
In our case we cannot be explicit, since passwords should not be passed in the open, and since .pgpass is a convenient way to handle passwords and it also contains the rest of the connection arguments we need to make a connection string  I see no reasons not to use it.

Reply:
I am not asking to pass the password as a part of connection string.
We just need to pass two arguments:
1. connection string (without password)
2. location of .pgpass

It will simplify a lot of things and will allow to get rid from ~75 lines of code parsing .pgpass...